### PR TITLE
Support macOS Aarch64 ABI for compiled wrappers.

### PR DESCRIPTION
**Pull request by gonzalolarralde**
This PR adds support for Apple's aarch64 C ABI, that requires stack arguments to be sized according to its data type, and considering padding properly.

Reference: https://developer.apple.com/documentation/xcode/writing_arm64_code_for_apple_platforms

/cc @AntonKozlov

Ref: JDK-8253818

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/aarch64-port.git pull/3/head:pull/3` \
`$ git checkout pull/3`

Update a local copy of the PR: \
`$ git checkout pull/3` \
`$ git pull https://git.openjdk.org/aarch64-port.git pull/3/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3`

View PR using the GUI difftool: \
`$ git pr show -t 3`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/aarch64-port/pull/3.diff">https://git.openjdk.org/aarch64-port/pull/3.diff</a>

</details>

### CODE REVIEW

**Review by VladimirKempik — CHANGES_REQUESTED** [review]
Hello.
I think you need to get rid of macros
>handle_padding(bytes_used, 1);
>advance_offset(bytes_used, 1);

you just put platform dependent size of type on stack in generic code

handle_padding  can be replaced with openjdk's align_up() function
 and advance_offset is simple bytes_used += size of type on stack;

size_of_type_on_stack can be made under #ifdefs in the beginning of the file

check example in our interp patch - https://github.com/openjdk/aarch64-port/commit/368f06e1cb965a68997cad0722aabcbe1068aad7


```
    _stack_offset = align_up(_stack_offset, nativeDoubleSpace);
    __ ldr(r0, src);
    __ str(r0, Address(to(), _stack_offset));
    _stack_offset += nativeDoubleSpace;
```

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 1]
```diff
@@ -56,6 +56,42 @@
...
+#  define advance_offset(bytes_used, size) { bytes_used += size; }
+#else
+#  define handle_padding(bytes_used, size)
+#  define advance_offset(bytes_used, size) { bytes_used += 8; }
+#endif
+
+// abi-dependant move for required types
+#ifdef __APPLE__
+#  define abi_dep_word_move(masm, src, dst, size) move_merge_and_shift(masm, src, dst, size);
+#  define abi_dep_float_move(masm, src, dst, size) move_merge_and_shift(masm, src, dst, size);
```
Comment (AntonKozlov) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:89]: `move_merge_and_shift` is unable to load to a float register (i.e. in stack->reg move), only to a general purpose one. I wonder why this was not caught by tests, but it looks like a bug

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 2]
```diff
@@ -770,6 +806,10 @@ AdapterHandlerEntry* SharedRuntime::generate_i2c2i_adapters(MacroAssembler *masm
   return AdapterHandlerLibrary::new_entry(fingerprint, i2c_entry, c2i_entry, c2i_unverified_entry, c2i_no_clinit_check_entry);
 }
 
+// On type size dependant ABIs types smaller than 8 bytes will use the second word
+// to store the position relative to the doubleword where the value needs to be
+// written. This allows padding to be pre-calculated. As a result of this it is
+// to be expected that some contiguous arguments share the same stack offset.
 int SharedRuntime::c_calling_convention(const BasicType *sig_bt,
```
Comment (AntonKozlov) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:813]: `c_calling_convention` is used by the shared VM code, which may be confused by non-VMRegs in regs. Although I assume shared code asks convention for short signatures only, some form of assert will be useful anyway, so we'll be sure shared code won't produce unexpected result.

For example, it is possible to make a private c_calling_convention, which will be called only from the aarch64 code, which can deal with unaliged offsets on stack. And to create another "public" c_calling_convention that does not try to encode anything unusual in regs. As a side benefit, the first "private" function can encode unaligned offsets in a more straightforward way, like a a separate bitmap or array argument.

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 3]
```diff
@@ -856,7 +906,10 @@ int SharedRuntime::c_calling_convention(const BasicType *sig_bt,
       }
     }
 
-  return stk_args;
+  // Ensure padding when finished parsing arguments
+  bytes_used += bytes_used % 8;
```
Comment (AntonKozlov) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:910]: `align_up` will be a bit more straightforward

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 4]
```diff
@@ -856,7 +906,10 @@ int SharedRuntime::c_calling_convention(const BasicType *sig_bt,
...
+
+  return bytes_used / 4;
```
Comment (AntonKozlov) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:912]: Please consider using VMRegImpl::stack_slot_size (4, a fixed size slot equal `sizeof(jint)`), wordSize (8, native word size) and VMRegImpl::slots_per_word (wordSize / stack_slot_size)

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 5]
```diff
@@ -885,6 +938,56 @@ static void move32_64(MacroAssembler* masm, VMRegPair src, VMRegPair dst) {
...
+  if (dst.first()->is_stack()) {
+      // FIXME: using second VMReg as bit offset to precalculate padding
+      int bit_offset = register_to_offset(dst.second()) * 8;
+      assert(bit_size + bit_offset <= 64, "can't offset outside 64 bits.");
+
+      // load source data into rscratch1 indistintevely of where it's coming from
+      if (src.first()->is_stack()) {
+        __ ldr(rscratch1, Address(rfp, reg2offset_in(src.first())));
+      } else {
+        __ mov(rscratch1, src.first()->as_Register());
```
Comment (AntonKozlov) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:959]: In this case, it's better operate directly on the `src.first()->as_Register()`

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 6]
```diff
@@ -885,6 +938,56 @@ static void move32_64(MacroAssembler* masm, VMRegPair src, VMRegPair dst) {
...
+        __ ror(rscratch2, rscratch2, bit_offset);
+      }
+      __ lsr(rscratch2, rscratch2, bit_size);
+      __ ror(rscratch2, rscratch2, 64 - bit_size - bit_offset);
+
+      // combine
+      __ orr(rscratch1, rscratch1, rscratch2);
+
+      // save
+      __ str(rscratch1, Address(sp, reg2offset_out(dst.first())));
```
Comment (AntonKozlov) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:980]: Can't this be replaced with just `lsr` and `strw/strh/strb`?

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 7]
```diff
@@ -1705,6 +1808,21 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
...
+        int_args++;
+        break;
+      case T_INT:
+        abi_dep_word_move(masm, in_regs[i], out_regs[c_arg], 32);
+        int_args++;
+        break;
+
       default:
         move32_64(masm, in_regs[i], out_regs[c_arg]);
         int_args++;
```
Comment (AntonKozlov) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:1828]: It seems we have processed all necessary types above, shouldn't the default case be `ShouldNotReachHere()`?

**Review by AntonKozlov — CHANGES_REQUESTED** [review]
Hi, I think there is a correctness issue that I would like to discuss

**Comment by gonzalolarralde** [comment]
Thanks for the comments @AntonKozlov, I'll try to rebuild using the latest changes on the branch that add support for MAP_JIT and get back with the feedback addressed. Due to the lack of a workaround my build + test runs were blocked, but now that those patches are public that should unblock me.

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp (hunk 1 above)
Comment (gonzalolarralde) [reply src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:89]: ah, I overlooked that. it seems that test cases might not be causing a displacement of float parameters, therefore never making this evaluate `true`: `(src.first() != dst.first())`. I will try to force the error and add a test if I confirm this to be the case.

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp (hunk 2 above)
Comment (gonzalolarralde) [reply src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:813]: that makes sense, will address this as you're describing. thanks!

**Pull request closed without merging**